### ISSUE TITLE
Documentation for extending ramda with plugins

### ISSUE THREAD

**Issue by bobiblazeski** [ramda/ramda#765]
I have some functionality that isn't in the spirit of ramda, could you write documentation how add my  functionality in a plugin for ramda.

**Comment by CrossEye** [comment]
It's really never come up as an issue, perhaps because it's easier than you might think.

Just think of the exposed Ramda `R` variable as a namespace you can attach functions to, and you'll be fine:

    R.myNewFunction = function( /* params */ ) { /* ... */ };

There is no current mechanism by which your code can be bundled with Ramda except by downloading the source, adding your functions, and rebuilding.  That's not hard, but it would probably make it confusing for others looking at your source.

But in general, is there a compelling need to hook your custom code to Ramda's namespace?  You could just as easily attach this as

    myApp.myNewFunction = function( /* params */ ) { /* ... */ };

It's not that I object; if you want to attach your function to Ramda, that's fine.  I'm just not sure I understand why.

**Comment by bobiblazeski** [comment]
I have some utilities that I use over several projects, and when I prototype or write throwaway projects I just want one file to download for quick start. Probably you're right about R, I'll manually merge my functions with ramdas into different namespace.

**Comment by bobiblazeski** [comment]
@CrossEye  Here is the documentation of the functionality that I've added into ramda https://github.com/bobiblazeski/Q and here's example of using it  in my latest project https://github.com/bobiblazeski/cristobal

**Comment by CrossEye** [comment]
Ok, let's see how tired my fingers get on this tablet.  (I'm a ten-finger typist.  This little keyboard may kill me.  I may stop part-way through.)

While there's some interesting stuff in there, there are only two functions that I find compellingly different from Ramda equivalents, `assembler` and `innerJoin`.  Below I detail Ramda equivalents for almost all of them.  But they are often not exact equivalents, as Ramda's philosophy avoids overloading parameters with different types and avoids optional parameters.  Many of these functions are conceptually much closer to Underscore/lodash in feel.  Those libraries are fine with the overloading and optional parameters.

The only ones I didn't try to write an equivalent for were `amend` and `innerJoin`.  They're not so easy to do in the Ramda points-free style I adopted for the others.  I also skipped `lay` and `taper` for philosophical reasons.  Many of the remaining ones are trivial.  But some are fairly interesting, especially the points-free implementation of `assembler`.

    var abate = R.useWith(R.map, R.I, R.values);
    abate(function(d){ return [-d,d];},{ a:4,b:9,c:16}); //=> [[-4, 4], [9, 9], [16, 16]]
    var abate = R.pipe(R.useWith(R.map, R.I, R.values), R.flatten);
    abate(function(d){ return [-d,d];},{ a:4,b:9,c:16}); //=> [-4, 4, 9, 9, 16, 16]



    // amend probably not easily points-free in Ramda; optional parameters difficult to do.



    var assembler = R.curryN(2, R.pipe(R.unapply(R.I), R.converge(R.zip, R.head, R.tail), R.fromPairs));
    var trend = assembler(["year", "amount", "name"]);
    trend(1937,100, 'foo'); //=> { year: 1937, amount: 100, name: "foo" }
    assembler(["year", "amount", "name"], 1937, 100, 'foo'); //=> { year: 1937, amount: 100, name: "foo" }



    var collect = R.pipe(R.mapObj, R.values);
    collect(Math.sqrt, {a: 1, b: 4, c: 9}); //=> [1, 2, 3]



    //curtail == abate?



    var division = R.flip(R.divide);
    division(100, 71) //=> 0.71
    R.map(division(16),[1,2,4,8,16]); //=> [0.0625, 0.125, 0.25, 0.5, 1]
    // Does not match Q.division documentation which reads [16,8,4,2,1], but probably shouldn't



    var group = R.groupBy;
    // Not quite the same as Q.group does parameter magic and turns a string parameter into `prop(str)`;
    // Ramda never does that, preferring to be explicit in such matters.
    R.groupBy(function(num) { return Math.floor(num); },[4.2, 6.1, 6.4]); 
    //=>( { '4': [4.2], '6': [6.1, 6.4] })



    // innerJoin: optional parameters, multiple types, certainly not within Ramda philosophy.



    // very similar to `propOr`, but with different parameter order and hand-holding on null/undefined objects
    var isNull = R.curryN(3, function(obj, name, defaultVal) {return R.propOr(defaultVal, name, obj || {});});
    isNull({a:1},"a",2) // => 1
    isNull({b:1},"a",2) // => 2
    isNull(null,"a",3) // => 3



    // lay:  see http://fr.umio.us/ranging-near-and-far/ for my argument against this version of `range`



    var least = R.minBy;
    // Not quite the same as Q.least does parameter magic and turns a string parameter into `prop(str)`;
    function cmp(obj) { return obj.x; } var a = {x: 1}, b = {x: 2}, c = {x: 3}; least(cmp, [a, b, c]); //=> {x: 1}
    // could also be written:
    least(prop('x'), [a, b, c]); //=> {x: 1}



    var minus = R.flip(R.subtract);
    minus(8, 10); //=> 2
    R.map(minus(10), [10,20,30]); //=> [0,10,20]



    var most = R.maxBy;
    // Not quite the same as Q.most does parameter magic and turns a string parameter into `prop(str)`;
    function cmp(obj) { return obj.x; } var a = {x: 1}, b = {x: 2}, c = {x: 3}; most(cmp, [a, b, c]); //=> {x: 3}
    // could also be written:
    most(prop('x'), [a, b, c]); //=> {x: 3}



    var order = Q.sortBy;
    // Not quite the same as Q.order does parameter magic and turns a string parameter into `prop(str)`;



    var remainder = R.flip(R.modulus);
    remainder(3,17); //=> 2 // JS behavior:
    remainder(3,-17); //=> -2
    remainder(17, -3); //=> 2



    // docs say list or object, but code seems only list.
    var sift = R.filter
    var isEven = function(n) { return n % 2 === 0; }; 
    sift(isEven, [1, 2, 3, 4]); //=> [2, 4]



    var single = R.find;
    // Not quite the same as Q.single does parameter magic and turns an object parameter into `where(obj)`;
    var xs = [{a: 1}, {a: 2}, {a: 3}]; 
    single(function(d){ return d.a == 2;}, xs); //=> {a: 2} 
    single(R.where({a: 2}), xs); //=> {a: 2}
    single(R.where({a: 4}), xs); //=> undefined


    var sumBy = R.pipe(R.map, R.sum);
    // Not quite the same as Q.sumBy does parameter magic and turns a string parameter into `prop(str)`;
    sumBy(R.prop("a"),[{a:2},{a:4},{a:6},{a:8},{a:100},{a:1}]); //=> 121
    sumBy(function(d){ return 2* d.a;},[{a:2},{a:4},{a:6},{a:8},{a:100},{a:1}]); //=>242
    sumBy(R.I,[2,4,6,8,100,100]); //=> 220



    // taper: See any of these links for discussions on folding objects:

    // https://github.com/ramda/ramda/issues/180
    // https://github.com/ramda/ramda/pull/257, especially
    // https://github.com/ramda/ramda/pull/257#issuecomment-52532396
    // https://github.com/ramda/ramda/pull/364
    // https://github.com/ramda/ramda/issues/546
    // https://github.com/ramda/ramda/pull/625

**Comment by bobiblazeski** [comment]
Very nice post, point free style is great I enjoyed doing it in APL family,  http://archive.vector.org.uk/art10500180 in JavaScript I'm mostly working with flat arrays of objects so parameter dispatch is very important to me.

**Comment by CrossEye** [comment]
@bobiblazeski: 

> I'm mostly working with flat arrays of objects so parameter dispatch is very important to me.

Flat arrays of object is what we consider Ramda's strong suit.  :smile:

But I'm curious as to which of two possible meanings you attach to "parameter dispatch".  Ramda has some objections to both, but for very different reasons.

  * One possibility is dispatching on function arity, that is, the number of parameters passed to the function.  This one is common in Javascript, and I personally don't have any major philosophical objections to it.  But there is an overwhelming practical problem with it for Ramda.  One of Ramda's key selling points is the auto-curried functions.  But to automatically curry a function, it's `length` property must be accurate.  This won't work if you want to be able to call it with various different numbers of parameters.

  * The other possibility is dispatching on argument types.  There is no practical issue with this, and, in fact, Ramda does have a dynamic dispatch mechanism for some key functions.  But for a more general type-checking, I hold strong objections.  Ramda was meant to be a simple library, in the sense described by the now classic [talk by Rich Hickey][rh], meaning that distinct concerns are never allowed to become intertwined.  A function should do one thing, and if we find ourselves branching on the types of the parameters, our function must be trying to do more than one thing; it's intertwining what should be the work of two separate functions.

So Ramda objects to both, for very different reasons.  Is one of these what you meant, and if so, can you explain why working with flat arrays of objects makes parameter dispatch so important to you?

The chief distinction I see between your functions and the Ramda style can be described by

    var stooges = [{name: 'moe', age: 41}, {name: 'larry', age: 53}, {name: 'curly', age: 39}];
    var decade = function(stooge) {return Math.round(stooge.age / 10, 0);}
    // Work the same
    Q.order(decade, stooges);
    R.sortBy(decade, stooges);
    // Different here
    Q.order('name', stooges); // shortcut available
    R.sortBy(R.prop('name'), stooges); // no shortcut offered

Is that sort of short-cut significant to your work?  Enough to complicate your implementation and to make your function signatures that much more complex?  The answer for Ramda has been no.  Obviously you have to make your own choice, but I would argue strongly that simplicity trumps easiness here.

  [rh]: http://www.infoq.com/presentations/Simple-Made-Easy

**Comment by bobiblazeski** [comment]
> Flat arrays of object is what we consider Ramda's strong suit.  :smile:

:smile:

> So Ramda objects to both, for very different reasons. Is one of these what you meant, and if so, can you explain why working with flat arrays of objects makes parameter dispatch so important to you?

From what I see from my code lack of dispatching on function arity is the right choice, the only function is lay and from my use it seems that is rarely used, so range works well. 

Dispatching on argument type unfortunately is deal breaker.In 98% of cases I have some array of objects that are one level deep and I want to do some transformations on it.  Its very convenient for me express what I want as fields.  It feels like SQL working with tables. 

> Is that sort of short-cut significant to your work? 

Its very convenient, http://dilbert.com/strip/2004-11-22 passing strings and objects represent the majority of  my use , functions are rare fallback. My number one feature in my wishlist is to extract full mongodb filtering syntax from   https://github.com/crcn/sift.js or https://github.com/kofrasa/mingo and add it.  If you grep over https://github.com/bobiblazeski/cristobal you'll see that. 

Loved that talk about Rich Hickey, he's a genious but unfortunately from talking with my lisper friends it seems that  Scala won and Clojure lost.  I've seen too many failed geniuses to know that understanding target market is very important.  Mark Tarver comes to mind first http://www.shenlanguage.org/ its not enough make something great in order to cross the chasm and go mainstream you need pragmatist who will adopt it.

If you want mainstream adoption for ramda you have to understand what market wants. If you think I represent   small niche ignore my suggestions and continue with your path. On the other hand if my needs represent large chunk of people  who might be interested to switch to more functional style, then somebody else will win and ramda will vanish into obscurity. I believe there is a nice opening for ramda in the d3 visualization market.    http://slobodanblazeski.blogspot.com/2010/05/becoming-popular.html

In the end time will tell.  Whether ramda, lodash-fp or something else something will fill the void.

**Comment by buzzdecafe** [comment]
> If you want mainstream adoption for ramda you have to understand what market wants.

This lib started as a personal project to study a particular style of coding and pursue it as far as we could. Somewhere along the way, other people got interested in the project too. That has certainly been fun, but "going mainstream" has never been a goal of mine.

Ramda is something I do in my free time for my own pleasure and edification. I can't speak for my collaborators, but "what the market wants" is *way down* my list of priorities. To use your terms, I prefer the Clojure path to the Scala path.

**Comment by CrossEye** [comment]
> I have some array of objects that are one level deep and I want to do some transformations on it. Its very convenient for me express what I want as fields. It feels like SQL working with tables.

But that doesn't imply dispatching on type.  It would be very easy to write a function

    // sortByField:: String -> [a] -> [a]
    var sortByField = R.useWith(R.sortBy, R.prop, R.I);
    sortByField('name', stooges);

and only use 

   R.sortBy(decade, stooges);

when you needed the broader power of a full function.  This doesn't dispatch on the type or run afoul the Ramda philosophy of simplicity.


> Loved that talk about Rich Hickey, he's a genious but unfortunately from talking with my lisper friends it seems that Scala won and Clojure lost. 

I haven't seen it that way.  Scala is older, and has slightly more job postings, but both are small players in the overall community.  Clojure, like all LISPs is not a purely functional language, but it can be used for truly functional programming; that's it's main mode, perhaps.  Scala is more promiscuous in its choice of paradigms.  But I see no signs that either is winning or losing at the expense of the other.  (I'm personally more interested in Haskell, but that's a different story.)


> If you want mainstream adoption for ramda you have to understand what market wants.

You've got the cart and the horse mixed up here.  I'm working on this, as @buzzdecafe said, for "my own pleasure and edification."  I found that I wanted to work in a certain manner, and the libraries that were available did not cater to manner.  So I started to write one.  @buzzdecafe and I, and eventually a large number of other contributors, have built this up in order to be able to _use_ it.  I can't deny that I've been gratified to find that it's filled the needs of many others, too.  But it's the ability to work in that manner that's of paramount importance; whatever popularity it has must flow from that.  The founder of lodash has [described his goals][jd] in terms of popularity.  I'd prefer to leave that to lodash and stick to our underlying principles.


  [jd]: https://github.com/ramda/ramda/issues/452#issuecomment-59665088

**Comment by GingerPlusPlus** [comment]
> ```js
> var sumBy = R.pipe(R.map, R.sum);
> ```

@CrossEye: this unnecessarily creates temporary array, unlike `_reduce` based implementation.

Also, `sumBy` has an interesting property: when passed a predicate, it'll return number of elements that satisfy it.

**Comment by CrossEye** [comment]
@GingerPlusPlus : Is your concern one of memory consumption?  Of speed?  Of cleanliness?

I will take clean code for as long as I can, optimizing for performance only when I can demonstrate that the code in question is causing a serious problem.  If I needed to do something here to avoid the intermediate objects, I might try the same code with a transducer to see if that's enough before I went all the way to rewriting with `reduce`. 

And by the way, what's with the commenting on a minor part of a four-year old comment to a long-closed issue?  :smile:
